src/service/video-manager-impl.js

```diff
@@ -959,6 +932,369 @@ class VideoEntry {
+  }
+
+  /** @private */
+  onRotation_() {
+    const root = this.ampdoc_.getRootNode();
+    if (this.isLandscape_()) {
+      const centeredVideoSelector = `.${CENTERED_PORTRAIT}.${BLESSED}`;
+      const video = root.querySelector(centeredVideoSelector);
+      if (!video) {
+        return;
+      }
+      const id = this.getId_(video);
+      this.enter_(dev().assert(this.entries_[id]));
+      return;
+    }
+
+    // Transfer visibility calculated in landscape to match.
+    // Rationale is that if the element is fully in view
+    // in landscape, it's most likely it will still be fully visible after
+    // rotation.
+    // This prevents a bug where videos blessed in landscape won't enter
+    // fullscreen after rotating back.
+    const videoInLandscapeRange = root.querySelector(CENTERED_LANDSCAPE);
+    if (videoInLandscapeRange) {
+      videoInLandscapeRange.classList.remove(CENTERED_LANDSCAPE);
+      videoInLandscapeRange.classList.add(CENTERED_PORTRAIT);
+    }
+
+    if (!this.currently_) {
+      return;
+    }
+    this.exit_(this.currently_);
+  }
+
+  /**
+   * @param {!R82FullscreenEntryDef} entry
+   * @private
+   */
+  enter_(entry) {
+    const {video, videoEntry} = entry;
+
+    if (!videoEntry.isPlaying()) {
+      return;
+    }
+
+    const platform = Services.platformFor(this.win_);
+
+    this.currently_ = entry;
+
+    if (platform.isAndroid() && platform.isChrome()) {
+      // Chrome on Android somehow knows what we're doing and executes a nice
+      // transition by default. Delegating to browser.
+      video.fullscreenEnter();
+      return;
+    }
+
+    this.scrollIntoIfNotVisible_(video)
+        .then(() => video.fullscreenEnter());
+  }
+
+  /**
+   * @param {!R82FullscreenEntryDef} entry
+   * @private
+   */
+  exit_(entry) {
+    this.currently_ = null;
+
+    const {video} = entry;
+
+    this.scrollIntoIfNotVisible_(video)
+        .then(() => video.fullscreenExit());
+  }
+
+  /**
+   * Scrolls to a video if it's not in view.
+   * @param {!../video-interface.VideoInterface} video
+   * @private
+   */
+  scrollIntoIfNotVisible_(video) {
+    const {element} = video;
+    const videoForVsync = /** @type {!../base-element.BaseElement} */ (video);
+
+    const viewport = this.getViewport_();
+
+    const duration = 300;
+    const curve = 'ease-in';
+    let pos; // 'top' or 'bottom'
+
+    let shouldScroll = true;
+
+    return this.onceOrientationChanges_().then(() =>
+      new Promise(resolve => {
+        videoForVsync.measureMutateElement(() => {
+          const {top, bottom} = element./*OK*/getBoundingClientRect();
+          const vh = viewport.getSize().height;
+          if (top >= 0 && bottom <= vh) {
+            shouldScroll = false;
+            return;
+          }
+          pos = (bottom > vh) ? 'bottom' : 'top';
+        }, () => {
+          if (!shouldScroll) {
+            resolve();
+            return;
+          }
+          this.getViewport_()
```

Comment:
would scrolling to `center` work? Like to avoid the calculations done here to figure out top vs bottom if possible